### DISCUSSION TITLE
How do I format embedded JSON strings

### DISCUSSION THREAD

**Discussion by ChipWolf** [mikefarah/yq#2220] [Q&A]
**Describe the bug**  
When attempting to apply `style="literal"` to JSON string values in a YAML file using `yq`, the transformation unexpectedly affects **all string values**, not just the targeted JSON strings, even when using `select()` to distinguish between JSON and non-JSON values.

This behavior makes it difficult to pretty-print JSON strings without inadvertently modifying other strings in the document.

Version of yq: 4.44.5
Operating system: linux
Installed via: binary release

---

**Input Yaml**  
data.yml:
```yaml
json_value: '{"key1": "value1", "key2": [1, 2, 3]}'
non_json_value: "This is a plain string."
```

---

**Command**  
The command I ran:
```bash
yq eval-all '
  .. |= (select(tag == "!!str" and test("^{.*}$")) |= (fromjson | tojson | . style="literal"))
' data.yml
```

---

**Actual behavior**
```yaml
json_value: "{\n  \"key1\": \"value1\",\n  \"key2\": [\n    1,\n    2,\n    3\n  ]\n}"
non_json_value: "This is a plain string."
```

---

**Expected behavior**
```yaml
json_value: |-
  {
    "key1": "value1",
    "key2": [
      1,
      2,
      3
    ]
  }
non_json_value: "This is a plain string."
```

---

**Workaround**  
A two-step `yq` command works as expected:
```bash
yq eval-all '
  .. |= (select(tag == "!!str" and test("^{.*}$")) |= (fromjson | tojson | . style="literal"))
' data.yml | yq eval-all '
  .. |= (select(tag == "!!str" and (test("^{.*}$") | not)) |= . style="")
'
```

This outputs:
```yaml
json_value: |-
  {
    "key1": "value1",
    "key2": [
      1,
      2,
      3
    ]
  }
non_json_value: This is a plain string.
```

---

**Additional context**  
Is it possible to achieve this in a single `eval-all` command? Or is this a limitation of how `style="literal"` and `select()` interact within a single evaluation pass? If this is intended behavior, I’d love to understand more about why it happens.

I've also attempted to use `with()`, with similar results.

**Comment by mikefarah** [comment]
Hey - you don't quite have the right expression; I think you're mixing up `|=` which updates value w.r.t themselves, and `|` which pipes the output of one filter into the next. You can do it like this:

```
yq 'with(.. | select(tag == "!!str" and test("^{.*}$")); 
    . |= (fromjson | . style="literal" | tojson ) | 
    . style = "literal"
)' data1.yaml
```

Explanation:
- We want to match all the strings that appear to be json; so we match all nodes then filter using select: `.. | select(tag == "!!str" and test("^{.*}$")`
- There are two updates we need to do, one to the value (format the stringified JSON) and another to the string style (to make it a block instead of single quotes. So we'll use `with` to allow us to make both those updates
- `. |= exp` updates the node value w.r.t itself, we'll use that to decode the json, reformat using style and re-encode back into a string
- then we can also apply a literal style to the string node itself, to make it a block

Hope that made sense!
